src/main/java/io/cryostat/rules/MatchExpressionManager.java

```diff
@@ -0,0 +1,211 @@
+            MatchExpressionValidator matchExpressionValidator,
+            Lazy<MatchExpressionEvaluator> matchExpressionEvaluator,
+            PlatformClient platformClient,
+            MatchExpressionDao dao,
+            Gson gson,
+            Logger logger) {
+        this.matchExpressionValidator = matchExpressionValidator;
+        this.matchExpressionEvaluator = matchExpressionEvaluator;
+        this.platformClient = platformClient;
+        this.dao = dao;
+        this.gson = gson;
+        this.logger = logger;
+    }
+
+    public int addMatchExpression(String matchExpression)
+            throws MatchExpressionValidationException {
+        matchExpressionValidator.validate(matchExpression);
+        MatchExpression expression = dao.save(new MatchExpression(matchExpression));
+        return expression.getId();
+    }
+
+    public Optional<MatchExpression> get(int id) {
+        return dao.get(id);
+    }
+
+    public List<MatchExpression> getAll() {
+        return dao.getAll();
+    }
+
+    public boolean delete(int id) {
+        return dao.delete(id);
+    }
+
+    public Set<ServiceRef> resolveMatchingTargets(int id) {
+        Optional<MatchExpression> matchExpression = dao.get(id);
+        if (matchExpression.isEmpty()) {
+            return Set.of();
+        }
+        return resolveMatchingTargets(matchExpression.get());
+    }
+
+    public Set<ServiceRef> resolveMatchingTargets(MatchExpression expr) {
+        Set<ServiceRef> matchedTargets = new HashSet<>();
+        for (ServiceRef target : platformClient.listDiscoverableServices()) {
+            try {
+                if (matchExpressionEvaluator.get().applies(expr.getMatchExpression(), target)) {
+                    matchedTargets.add(target);
+                }
+            } catch (ScriptException e) {
+                logger.error(e);
+                break;
+            }
+        }
+        return matchedTargets;
+    }
+
+    public Set<ServiceRef> resolveMatchingTargets(
+            MatchExpression expr, Collection<ServiceRef> targets) {
+        return resolveMatchingTargets(expr.getMatchExpression(), targets);
+    }
+
+    public Set<ServiceRef> resolveMatchingTargets(String expr, Collection<ServiceRef> targets) {
+        Set<ServiceRef> matchedTargets = new HashSet<>();
+        for (ServiceRef target :
+                platformClient.listDiscoverableServices().stream()
+                        .filter(t -> targets.contains(t))
+                        .toList()) {
+            try {
+                if (matchExpressionEvaluator.get().applies(expr, target)) {
+                    matchedTargets.add(target);
+                }
+            } catch (ScriptException e) {
+                logger.error(e);
+                break;
+            }
+        }
+        return matchedTargets;
+    }
+
+    public List<ServiceRef> parseTargets(String targets) {
+        Objects.requireNonNull(targets, "Targets must not be null");
```

Comment:
Is this still worth keeping as a member of the Manager class? Seems like it could just be inlined into the associated Handler.